### PR TITLE
feat: add oauth2 token exchange

### PR DESCRIPTION
**Pull request by code-asher**
Re-opened since the original got closed by stale bot and will not let me re-open after force-pushing!

`Rename token hashed_secret to refresh_hash` and later are new since that PR was last reviewed.  The main thing is that a secret prefix was added.

https://github.com/coder/coder/pull/11778

### CODE REVIEW

File: enterprise/coderd/oauth2.go [hunk 1]
```diff
@@ -208,10 +208,14 @@ func (api *API) postOAuth2ProviderAppSecret(rw http.ResponseWriter, r *http.Requ
 		})
 		return
 	}
+	// TODO: Currently unused.
+	prefix, _ := cryptorand.String(40)
```
Comment (code-asher) [inline enterprise/coderd/oauth2.go:212]: Basically a stop-gap, this is replaced in the next PR.  I figure once this is approved I merged that one here, and then merge both into main together.

File: coderd/database/dbmem/dbmem.go [hunk 2]
```diff
@@ -1204,12 +1265,74 @@ func (q *FakeQuerier) DeleteOAuth2ProviderAppSecretByID(_ context.Context, id uu
...
+			// Delete from API keys.
+			var keys []database.APIKey
+			for _, key := range q.apiKeys {
+				if !slices.Contains(keyIDsToDelete, key.ID) {
+					keys = append(keys, key)
+				}
+			}
+			q.apiKeys = keys
+
 			return nil
```
Comment (Emyrk) [inline coderd/database/dbmem/dbmem.go:1290]: nesting all this logic in the for loop feels a bit difficult to read. Maybe it's just me, but this feels a bit cleaner imo.

```golang
	index := slices.IndexFunc(q.oauth2ProviderAppSecrets, func(secret database.OAuth2ProviderAppSecret) bool {
		return secret.ID == id
	})

	if index < 0 {
		return sql.ErrNoRows
	}

	q.oauth2ProviderAppSecrets[index] = q.oauth2ProviderAppSecrets[len(q.oauth2ProviderAppSecrets)-1]
	q.oauth2ProviderAppSecrets = q.oauth2ProviderAppSecrets[:len(q.oauth2ProviderAppSecrets)-1]
	// ... rest of delete ...

```

**Review by Emyrk — APPROVED** [review]

File: coderd/database/dbmem/dbmem.go [hunk 3]
```diff
@@ -1182,14 +1185,72 @@ func (q *FakeQuerier) DeleteOAuth2ProviderAppByID(_ context.Context, id uuid.UUI
...
+			}
+			q.oauth2ProviderAppTokens = tokens
+
+			// Delete from API keys.
+			var keys []database.APIKey
+			for _, key := range q.apiKeys {
+				if !slices.Contains(keyIDsToDelete, key.ID) {
+					keys = append(keys, key)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (code-asher) [reply coderd/database/dbmem/dbmem.go:1218]: Oh snap TIL.  Will use.

File: coderd/database/dbmem/dbmem.go (hunk 2 above)
Comment (code-asher) [reply coderd/database/dbmem/dbmem.go:1290]: You are right, this is much better.

Comment (code-asher) [reply coderd/database/dbmem/dbmem.go:1290]: I do this in a few places, going to update all of them along with using `DeleteFunc`!

**Pull request merged**